File: Java/BlockSort.java
public class BlockSort {
    public static void blockSort(int[] array) {
        int n = array.length;
        if (n <= 1) return;
        int max = array[0], min = array[0];
        for (int i = 1; i < n; i++) {
            if (array[i] > max) max = array[i];
            if (array[i] < min) min = array[i];
        }
        int range = max - min + 1;
        int[] count = new int[range];
        for (int i = 0; i < n; i++) {
            count[array[i] - min]++;
        }
        int index = 0;
        for (int i = 0; i < range; i++) {
            while (count[i]-- > 0) {
                array[index++] = i + min;
            }
        }
    }

    public static void main(String[] args) {
        int[] array = {4, 7, 2, 8, 3, 13, 1, 0};
        blockSort(array);
        for (int i : array) {
            System.out.print(i + " ");
        }
    }
}

File: PHP/block_sort.php
<?php

function blockSort(array $arr): array {
    $n = count($arr);
    if ($n <= 1) {
        return $arr;
    }

    $min = min($arr);
    $max = max($arr);

    $range = $max - $min + 1;
    $blockSize = ceil($range / $n);
    $blocks = array_fill(0, $n, []);

    foreach ($arr as $value) {
        $index = (int)(($value - $min) / $blockSize);
        $blocks[$index][] = $value;
    }

    $sortedArray = [];
    foreach ($blocks as $block) {
        sort($block);
        $sortedArray = array_merge($sortedArray, $block);
    }

    return $sortedArray;
}

$array = [29, 25, 3, 49, 9, 37, 21, 43, 0, -7, 3];
$sortedArray = blockSort($array);

echo "Sorted Array: " . implode(", ", $sortedArray) . PHP_EOL;
?>
